static/app/components/dropdownMenuV2.tsx

```diff
@@ -0,0 +1,301 @@
+   * This information is used for controlling keyboard events.
+   * See: modifiedMenuProps below.
+   */
+  const [hasFocus, setHasFocus] = useState(true);
+  useEffect(() => {
+    /**
+     * A submenu is a leaf when it does not contain any expanded submenu.
+     * This logically follows from the tree-like structure and single-selection
+     * nature of menus.
+     */
+    const isLeafSubmenu = !stateCollection.some(node => {
+      const isSection = node.hasChildNodes && !node.value.isSubmenu;
+      /**
+       * A submenu with key [key] is expanded if
+       * state.selectionManager.isSelected([key]) = true
+       */
+      return isSection
+        ? [...node.childNodes].some(child =>
+            state.selectionManager.isSelected(`${child.key}`)
+          )
+        : state.selectionManager.isSelected(`${node.key}`);
+    });
+    setHasFocus(isLeafSubmenu);
+  }, [state.selectionManager.selectedKeys]);
+  /**
+   * Menu props from useMenu, modified to disable keyboard events
+   * if the current menu does not have focus.
+   */
+  const modifiedMenuProps = {
+    ...menuProps,
+    ...(!hasFocus && {
+      onKeyUp: () => null,
+      onKeyDown: () => null,
+    }),
+  };
+
+  /**
+   * Render a single menu item
+   */
+  const renderItem = (node: Node<MenuItemProps>, isLastNode: boolean) => {
+    return (
+      <MenuItem
+        node={node}
+        isLastNode={isLastNode}
+        state={state}
+        onClose={closeRootMenu}
+        closeOnSelect={closeOnSelect}
+      />
+    );
+  };
+
+  /**
+   * Render a submenu whose trigger button is a menu item
+   */
+  const renderItemWithSubmenu = (node: Node<MenuItemProps>, isLastNode: boolean) => {
+    const trigger = ({props: submenuTriggerProps, ref: submenuTriggerRef}) => (
+      <MenuItem
+        ref={submenuTriggerRef}
+        renderAs="div"
+        node={node}
+        isLastNode={isLastNode}
+        state={state}
+        isSubmenuTrigger
+        {...submenuTriggerProps}
+      />
+    );
+
+    return (
+      <MenuControl
+        items={node.value.children as MenuItemProps[]}
+        trigger={trigger}
+        menuTitle={node.value.submenuTitle}
+        placement="right top"
+        offset={-4}
+        crossOffset={-8}
+        closeOnSelect={closeOnSelect}
+        isOpen={state.selectionManager.isSelected(node.key)}
+        isSubmenu
+        closeRootMenu={closeRootMenu}
+        closeCurrentSubmenu={() => state.selectionManager.clearSelection()}
+        renderWrapAs="li"
+      />
+    );
+  };
+
+  /**
+   * Render a collection of menu items
+   */
+  const renderCollection = (collection: Node<MenuItemProps>[]) =>
+    collection.map((node, i) => {
+      const isLastNode = collection.length - 1 === i;
+      const showSeparator =
+        !isLastNode && (node.type === 'section' || collection[i + 1]?.type === 'section');
+      const {separatorProps} = useSeparator({elementType: 'li'});
```

Comment:
Damn this library really is like, batteries included

Reply:
ikr!